maths/numerical_analysis/weierstrass_method.py

```diff
@@ -0,0 +1,85 @@
+from collections.abc import Callable
+
+import numpy as np
+
+
+def weierstrass_method(
+    polynomial: Callable[[np.ndarray], np.ndarray],
+    degree: int,
+    roots: np.ndarray | None = None,
+    max_iter: int = 100,
+) -> np.ndarray:
+    """
+    Approximates all complex roots of a polynomial using the
+    Weierstrass (Durand-Kerner) method.
+    Args:
+        polynomial: A function that takes a NumPy array of complex numbers and returns
+                    the polynomial values at those points.
+        degree: Degree of the polynomial (number of roots to find). Must be ≥ 1.
+        roots:  Optional initial guess as a NumPy array of complex numbers.
+                Must have length equal to 'degree'.
+                If None, perturbed complex roots of unity are used.
+        max_iter: Number of iterations to perform (default: 100).
+
+    Returns:
+        np.ndarray: Array of approximated complex roots.
+
+    Raises:
+        ValueError: If degree < 1, or if initial roots length doesn't match the degree.
+
+    Note:
+        - Root updates are clipped to prevent numerical overflow.
+
+    Example:
+        >>> import numpy as np
+        >>> def f(x): return x**2 - 1
+        >>> roots = weierstrass_method(f, 2)
+        >>> np.allclose(np.sort(roots), np.sort(np.array([1, -1])))
+        True
+
```

Comment:
Lets add some more tests:
```suggestion
        >>> def f(x): return x**3 - 6*x**2 + 11*x - 6
        >>> roots = weierstrass_method(f, 3)
        >>> np.allclose(np.sort(roots), np.array([1, 2, 3]))
        True
```